packages/iTwinUI-react/src/core/Table/actionHandlers/selectHandler.ts

```diff
@@ -88,6 +89,67 @@ export const onSingleSelectHandler = <T extends Record<string, unknown>>(
   return newState;
 };
 
+/**
+ * Handles selection when clicked on a row while shift key is pressed.
+ */
+export const onShiftSelectHandler = <T extends Record<string, unknown>>(
+  state: TableState<T>,
+  action: ActionType,
+  instance?: TableInstance<T>,
+  onSelect?: (
+    selectedData: T[] | undefined,
+    tableState?: TableState<T>,
+  ) => void,
+  isRowDisabled?: (rowData: T) => boolean,
+) => {
+  if (instance == null) {
+    return state;
+  }
+
+  let startIndex = instance.flatRows.findIndex(
+    (row) => row.id === state.lastSelectedRow,
+  );
+  let endIndex = instance.flatRows.findIndex((row) => row.id === action.id);
+
+  if (startIndex < 0) {
+    startIndex = 0;
+  }
+  if (endIndex < 0) {
+    endIndex = 0;
+  }
+
+  if (startIndex > endIndex) {
+    const temp = startIndex;
+    startIndex = endIndex;
+    endIndex = temp;
+  }
+
+  const selectedRowIds: Record<string, boolean> = {};
+
+  // 1. Select all rows between start and end
+  instance.flatRows
+    .slice(startIndex, endIndex + 1)
+    .forEach((r) => (selectedRowIds[r.id] = true));
+
+  // 2. Select all children of the last row (endIndex)
+  const handleRow = (row: Row<T>) => {
+    selectedRowIds[row.id] = true;
+    row.subRows.forEach((r) => handleRow(r));
+  };
+  handleRow(instance.flatRows[endIndex]);
```

Comment:
Why do you need to handle last row separately?

Reply:
That is because we are selecting the rows between the start and the end in the `flatRows` array. But the subRows of the end row comes after the end row in the `flatRows` array. So we need to manually select all the subRows of the end row.

E.g. If user selects 2 to 5, then 5.1 for example will not be selected even though 5 is selected

Reply:
Would it not be enough to call `handleRow` on every row from start to end?

Reply:
Can you show an example of how the rows are structured? Because I'm not sure I really understand why `2.2.1` is any different from `5.2.1` in terms of handling.

Reply:
Here is an example tree structure for selecting 2 to 4 (not 5; to keep the example small)

- 1
  - 1.1
  - ...
- **2 [START]**
  - 2.1
  - 2.2
    - 2.2.1
    - 2.2.2
    - 2.2.3
  - 2.3
  	- 2.3.1
  	- 2.3.2
- 3
  - 3.1
  - 3.2
    - 3.2.1
  - 3.3
  - 3.4
- 4
  - 4.1
  - **4.2 [END]**
   - 4.2.1
   - 4.2.2
   - 4.2.3
  - 4.3

Current approach: Treat the nested rows structure (rows has subrows...) as a flat array of just rows with no hierarchy (`flatRows`). Then iterate through this array from the start index to the end index and set each row as checked along the way.

> Why do you need to handle last row separately?

The above mentioned iteration stops at 4.2 since that is the last row (the row we clicked). But this loop does not select 4.2.1 - 4.2.3. This is why we have to handle the last row separately.

> Would it not be enough to call handleRow on every row from start to end?

If we started selection from 2.3 for example, then `handleRow(2.3)` would select itself, 2.3.1, and 2.3.2. The next row in `flatRows` is 2.3.1. Thus, `handleRow(2.3.1)` will select 2.3.1. Thus, 2.3.1 would be selected twice which is inefficient.

Reply:
> The above mentioned iteration stops at 4.2 since that is the last row (the row we clicked). But this loop does not select 4.2.1 - 4.2.3. This is why we have to handle the last row separately.

Ah, this finally clarifies it! Maybe add another line in the comment explaining this.

In general, comments explaining "why" are more useful than ones explaining "what".

Reply:
Currently the comment says _"Select all children of the last row (endIndex) Because the subRows of the lastRow have indices > endIndex and hence won't be selected by step 1"_

This is a bit misleading as we are not selecting "all" children and the reasoning also seems a bit convoluted. Would it be more accurate to say something like "_Handle last row separately because the user might select only a few of the subrows instead of all"_?

Reply:
Not quite. Actually, lines 136-140 is not because the user may select a few of the subrows instead of all rows. That code block was to indeed to select "all" the children (subRows and their subrows...) of the last row since they will not be selected by the loop in step 1.

But I do agree that is it a little wordy. So I changed it to the below lines. Do you think it is alright? Maybe I can make it shorter? Or a little clearer if it is ambiguous?

```tsx
// 2. Select all children of the last row (endIndex)
// Since lastRow's children come after endIndex + 1 (not selected in step 1)
```